File: map.py
from bitcoinrpc.authproxy import AuthServiceProxy
import simplejson as json
import time
import os, sys
from dotenv import load_dotenv

load_dotenv()

user = os.getenv("RPC_USER")
password = os.getenv("RPC_PASSWORD")
host = os.getenv("RPC_HOST")

rpc_connection = AuthServiceProxy("http://%s:%s@%s:8332"%(user, password,host))

def getBlock(height):
    blockhash = rpc_connection.getblockhash(height)
    block = rpc_connection.getblock(blockhash)
    return block

def getTx(txid):
    rawtx = rpc_connection.getrawtransaction(txid, True)
    return rawtx

def getOuts(rawtx):
    outs = []
    for vout in rawtx['vout']:
        if 'addresses' in vout['scriptPubKey']:
            for addr in vout['scriptPubKey']['addresses']:
                txin = {
                    'txidIn': rawtx['txid'],
                    'addr': addr,
                    'out': vout['n'],
                    'value': vout['value'],
                    'blockhash': rawtx['blockhash']
                }
                outs.append(txin)
    return outs

def getIns(tx):
    ins = []
    for vin in tx['vin']:
        if 'coinbase' not in vin:
            txSource = getTx(vin['txid'])
            voutSource = getOuts(txSource)

            for out in voutSource:
                if out['out'] == vin['vout']:
                    txout = {
                        'txidOut': tx['txid'],
                        'addr': out['addr'],
                        'in': vin['vout'],
                        'value': out['value'],
                        'blockhash': tx['blockhash']
                    }
                    ins.append(txout)
    return ins

def getFees(tx):
    valueOut = 0
    for tab in getOuts(tx):
        valueOut+=tab['value']

    valueIn = 0
    for tab in getIns(tx):
        valueIn+=tab['value']

    fees = valueIn-valueOut
    return fees

def txMapFunction(ins, outs, fees):
    result = {
        'ins': ins,
        'outs': outs,
        'fees': fees
    }
    return result

mapResult = []
start = time.time()
print("start", time.strftime("%H:%M:%S", time.localtime(start)))

#for x in range(1, 11):
#x = rpc_connection.getblockcount()
x = sys.argv[1]
block = getBlock(x)

#for each transaction
for i in block['tx']:
    rawtx = getTx(i)
    mapResult.append(txMapFunction(getIns(rawtx), getOuts(rawtx), getFees(rawtx)))

print("writing in file", time.strftime("%H:%M:%S", time.localtime(time.time())))
with open(str(x) + '.json', 'a') as outfile:
    json.dump(mapResult, outfile, indent=4)

end = time.time()
print("finish", time.strftime("%H:%M:%S", time.localtime(end)))

duration = end - start
print("duration", time.strftime("%H:%M:%S", time.gmtime(duration)))

File: functions/map.js
const decrypt = require('./decrypt')
const {getResult} = require('../rpc_config')

async function getBlock(hash){
    let mapTx = []
    var dataString = JSON.stringify({jsonrpc:"2.0",id:"curltext",method:"getblock",params:[`${hash}`]});
    const block = await getResult(dataString)
    let txids = block.tx
    let fini = 0;
    for (let i = 0; i < txids.length; i++) {
        let txid = txids[i]
        let tx = await getTx(txid)
        let voutTx = getOuts(tx)
        let vinTx = getIns(tx)

        await Promise.all([voutTx, vinTx]).then(values=>{
            mapTx.push({
                ins: values[0],
                outs: values[1]
            })
            if(++fini === txids.length){
                console.log('mapTx ok')
            }
        }).catch(e => {
            console.log('get block error', e)
        } )
    } 
    return new Promise(result=>{
        result(mapTx)
    })
}

function getTx(txid){
    return new Promise(result => {
        let dataString = JSON.stringify({jsonrpc:"2.0",id:"curltext",method:"getrawtransaction",params:[`${txid}`, true]});
        getResult(dataString).then(tx => {
            result(tx)
        })
    })
}

async function getOuts(rawtx){
    let outs = []
    let tab = rawtx.vout
    tab.forEach(vout => {
        if(vout.scriptPubKey.addresses != undefined){
            vout.scriptPubKey.addresses.forEach(addr => {
                let txin = {
                    txid: rawtx.txid,
                    address: addr,
                    n: vout.n,
                    value: vout.value,
                    blockhash: rawtx.blockhash,
                    blocktime: rawtx.time
                }
                outs.push(txin)
            });
        }else if(vout.scriptPubKey.type == "pubkey"){
            let txin = {
                txid: rawtx.txid,
                address: decrypt(vout),
                n: vout.n,
                value: vout.value,
                blockhash: rawtx.blockhash,
                blocktime: rawtx.time
            }
            //console.log('decrypted', rawtx.txid)
            outs.push(txin)
        }
    });
    return outs
}

async function getIns(rawtx){
    let ins = []
    let tab = rawtx.vin
    for (let k = 0; k < tab.length; k++) {
        let vin = tab[k];                       
        if(vin.coinbase == undefined){
            let txSource = await getTx(vin.txid)
            let outs = await getOuts(txSource)
            outs.forEach(out => {
                if(out.n == vin.vout){
                    let txout = {
                        txid: rawtx.txid,
                        address: out.address,
                        n: vin.vout,
                        value: out.value,
                        blockhash: rawtx.blockhash,
                        blocktime: rawtx.time
                    }
                    ins.push(txout)
                }                    
            })
        } else{
            getOuts(rawtx).then(outs=>{
                let value = 0
                outs.forEach(out => {
                    value += out.value                 
                })
    
                let txout = {
                    txid: rawtx.txid,
                    script: vin.coinbase,
                    n: vin.vout,
                    value: value,
                    blockhash: rawtx.blockhash,
                    blocktime: rawtx.time,
                    coinbase: true
                }
                //console.log('coinbase', rawtx.txid)
                ins.push(txout)
            })
        }
    }
    return ins
}

exports.getBlock = getBlock;